src/controllers/company.number.controller.ts

```diff
@@ -22,21 +22,27 @@ const preValidators = [
 const padCompanyNumber = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
   let companyNumber: string = req.body.companyNumber;
   if (/^([a-zA-Z]{2}?)/gm.test(companyNumber)) {
-    const leadingLetters = companyNumber.substring(0, 2);
-    let trailingChars = companyNumber.substring(2, companyNumber.length);
-    trailingChars = trailingChars.padStart(6, "0");
-    companyNumber = leadingLetters + trailingChars;
+    companyNumber = formatCompanyNumber(companyNumber, 2, 6);
+  } else if (/^([a-zA-Z]{1}?)/gm.test(companyNumber)) {
+    companyNumber = formatCompanyNumber(companyNumber, 1, 7);
   } else {
     companyNumber = companyNumber.padStart(8, "0");
   }
   req.body.companyNumber = companyNumber;
   return next();
 };
 
+const formatCompanyNumber = (companyNumber: string, leadPoint: number, padStart: number): string  => {
+  const leadingLetters = companyNumber.substring(0, leadPoint);
+  let trailingChars = companyNumber.substring(leadPoint, companyNumber.length);
+  trailingChars = trailingChars.padStart(padStart, "0");
+  return leadingLetters + trailingChars;
+};
+
 // validator middleware that checks for invalid characters in the input
 const postValidators = [
   check("companyNumber").blacklist(" ").escape().custom((value: string) => {
-    if (!/^([a-zA-Z]{2})?[0-9]{6,8}$/gm.test(value)) {
+    if (!/^([a-zA-Z]{1,2})?[0-9]{6,8}$/gm.test(value)) {
```

Comment:
Does this validator work if a 1 letter, 7 number string was passed to it?
Looks to me as if its expecting 'one or none' of 1 letter, or 2 letters.
Followed by either 6 numbers, or 8 numbers.
(My regex is rusty so I may be missing something here)

Reply:
It allows a 1 letter 6 number through which I don't think is correct.
H123456
you can test it on https://regex101.com/

Reply:
Seems it also lets through 'just 6 numbers' and '2 letters 8 numbers'

Reply:
try this : ^[0-9]{8}|([a-zA-Z]{1})[0-9]{7}|([a-zA-Z]{2})[0-9]{6}$

Reply:
This is based on existing pattern in extensions: 
if (!/^([a-zA-Z]{2})?[0-9]{6,8}$/gm.test(value)) {

so I thought we were aware of its shortcomings as this allows thing like:
HA12345678
345678

If it is of concern then we need to raise a ticket to fix this in extensions as well.

Reply:
^[0-9]{8}$|^([a-zA-Z]{1})[0-9]{7}$|^([a-zA-Z]{2})[0-9]{6}$
This appears to work? Adds start and end of string checks to each or section

Reply:
I understand why this ipassed previously if we look at the code as a whole the regex check happens in the post validator after we pad shart code such as "H123456" with 0s so this becomes H0123456 there is also a check to ensure that the string is 8 characters long.

Reply:
Should we update the validator anyway? I get that currently it all works as expected.
But if in future the padding method changes or doesn't work as expected. We want to ensure the validator would pick that up and fail it?

Reply:
can we update this regex to ^[0-9]{8}$|^([a-zA-Z]{1})[0-9]{7}$|^([a-zA-Z]{2})[0-9]{6}$. to make it more accurate

Reply:
The unit tests would alert us to that - there is test that looks at 6400 and expects 00006400, this would start to fail and prompt further changes. However, we want the code to be as robust as it can be so will change.

Reply:
Updated regex to this new pattern.